Test/tutorial/parser_combinators.dfy

```diff
@@ -0,0 +1,294 @@
+/// may need to be able to state that a recursive call to a parser can only
+/// happen with a larger index, for example.
+
+  type Parser<+T> = nat --> ParseResult<T>
+
+/// The following is a convenient trick when defining many functions that share
+/// an argument without having to pass that argument everywhere and without
+/// having to specify that all calls use the same value of that argument: we
+/// define an `Engine` to be a datatype that contains an `input` string, and
+/// members of the datatype can access it through the implicit `this` pointer
+/// (of course, they don't have to: some members below are `static`).
+
+  datatype Engine = Engine(input: string) {
+
+/// `Concat` takes two parsers and creates a new one: a lambda (captured in the
+/// type `Parser<(L, R)>`) that returns a pair of values, or `Failure`.  Because
+/// the input parsers are partial functions, so it their combination: the
+/// `requires` clauses on the lambda specify that the left parser may be called
+/// (`left.requires(pos)`) and that the right parser may be called, but only if
+/// the left one succeeds, and only on the position returned by it.
+
+    static function Concat0<L, R>(left: Parser<L>, right: Parser<R>)
+      : (p: Parser<(L, R)>)
+    {
+      (pos: nat)
+        requires left.requires(pos)
+        requires left(pos).Success? ==> right.requires(left(pos).pos)
+      =>
+        var (pos, l) :- left(pos);
+        var (pos, r) :- right(pos);
+        Success(pos, (l, r))
+    }
+
+/// The definition above is transparent: every time the verifier encounters it,
+/// it is allowed to peek inside the body of the function and use it for
+/// reasoning.  This will cause verification-performance issues down the line,
+/// so we use the following alternative definition instead:
+
+    static function {:opaque} Concat<L, R>(
+      left: Parser<L>,
+      right: Parser<R>
+    )
+      : (p: Parser<(L, R)>)
+      ensures forall pos: nat |
+        && left.requires(pos)
+        && (left(pos).Success? ==> right.requires(left(pos).pos))
+        :: p.requires(pos)
+    {
+      (pos: nat)
+        requires left.requires(pos)
+        requires left(pos).Success? ==> right.requires(left(pos).pos)
+      =>
+        var (pos, l) :- left(pos);
+        var (pos, r) :- right(pos);
+        Success(pos, (l, r))
+    }
+
+/// The two important things to note are (1) the fact that `Concat` is now
+/// marked `:opaque`, and (2) the fact that `Concat` now has an `ensures` clause
+/// of the following form:
+///
+/// ```dafny
+///   ensures forall input :: (precondition of the parser) ==> p.requires(input)
+/// ```
+///
+/// This does not say that the parser must accept all inputs: only that the
+/// condition `(precondition of the parser)` is sufficient to call the opaque
+/// function `p` that `Concat` returns.
+///
+/// With that in mind, we can define `Either` in the same way:
+
+    static function {:opaque} Either<L, R>(
```

Comment:
Is opaque strictly necessary?

Reply:
Here, yes, for performance.  I added some explanation above.